Task refactor

levb opened the pull request:
#### Summary
Draft PR, need to write up the summary - many changes.
The biggest is the addition of Tasks (tickets or shifts) and a significant refactoring. A lot of functional tests are now automated with commands.
This PR is still missing Autopilot, Guess, and Forecast capabilities to be at parity with the previous status quo, as well as many details. 

WIP - will edit later.

@iomodo @larkox - `master` has been reviewed by @crspeller and @cpoile, but they now moved on. I think the 2 of you are probably best fit for this plugin, apologies for a huge chunk.

levb commented on server/command/command.go:
```diff
@@ -173,25 +187,206 @@ func (c *Command) handleCommand(subcommands map[string]func([]string) (string, e
+			return nil, errors.New("`@username`'s expected")
+		}
+		arg = arg[1:]
+		user, err := c.SL.LoadMattermostUsername(arg)
```
That was my first name for it I think, but I thought it was "too Java". I'll think about it some more, somewhat agree that the current name is confusing.

levb commented on server/command/log.go:
```diff
@@ -6,23 +6,23 @@ package command
-	})
+	sc := c.SL.Config().StoredConfig
+	sc.AdminLogLevel = level
+	sc.AdminLogVerbose = verbose
```
This command was there for testing/demo purposes only; I'll rename it to `debug-log`; and yes one must provide all values at once, not designed to be user-friendly.

levb commented on server/command/rotation.go:
```diff
@@ -3,60 +3,20 @@
+import "github.com/mattermost/mattermost-plugin-solar-lottery/server/utils/md"
+
+func (c *Command) rotation(parameters []string) (md.MD, error) {
+	subcommands := map[string]func([]string) (md.MD, error){
```
Not sure I understand this comment, elaborate please?

larkox replied:
In `command.go` we have this function:
```
func (c *Command) commands() map[string]func([]string) (md.MD, error)
```
This function is at some point used as:
```
c.handleCommand(c.commands(), parameters)
```
Here is a similar pattern, but we declare the map inside the function, instead of creating a new function like commands() to create the map.

Probably for these cases with subcommands is enough as it is, but I am wondering if there would be any improvement from having a `rotationSubcommands()` function (and similar for any other command with subcommands).

levb replied:
Ah! https://github.com/levb/mattermost-plugin-solar-lottery/blob/lev-task-refactor/server/command/command_test.go#L107 uses it to `runCommand`, so it needed to be shareable. Returning a copy - just so that no need to worry about tests stepping onto each other.

levb commented on server/command/rotation_archive.go:
```diff
@@ -4,43 +4,40 @@
-func (c *Command) archiveRotation(parameters []string) (string, error) {
-	var rotationID, rotationName string
-	fs := newRotationFlagSet(&rotationID, &rotationName)
+func (c *Command) rotationArchive(parameters []string) (md.MD, error) {
```
Personally I don't like the "reverse" `rotationArchive` convention in the code - makes it difficult to read. But as a programmer who shies away from highly-automated code-browsing tools, I appreciate a meaningful sorting order.

In this project I have struggled to find a happy medium. My present thinking is that there's no way around naming files in reverse; in `command` there are so many little functions, hierarchical in nature, that after hesitation, I decided to rename them that way as well. I probably did not finish the job in the commands that weren't touched "naturally" by this refactoring. Need to make them consistent, I agree.

levb commented on server/command/rotation_param.go:
```diff
@@ -0,0 +1,122 @@
+	return c.rotationParamMinMax(false, parameters)
+}
+
+func (c *Command) rotationParamMinMax(min bool, parameters []string) (md.MD, error) {
```
1/5 I thought this was trivial, and a type would be an overkill.

levb commented on server/command/rotation_param.go:
```diff
@@ -0,0 +1,122 @@
+		return "", err
+	}
+
+	return c.normalOut(
```
I could, but that's the convention: all API functions return a value (`md.Markdowner` and/or `json.Marshaller`), and an error. I can move the `func()` into variable to reduce the indentation, but 1/5 prefer keeping the convention of `return c.normalOut(c.SL.someAPIfunction(...))`

levb commented on server/command/rotation_show.go:
```diff
@@ -3,25 +3,27 @@
+import "github.com/mattermost/mattermost-plugin-solar-lottery/server/utils/md"
+
+func (c *Command) rotationShow(parameters []string) (md.MD, error) {
+	fs := c.assureFS()
```
That's because all other c.fXXX() functions initialize c.fs lazily.

2/5 maybe replace pflag.Parse with a c.Parse and lazily init `fs` there. Will do.

levb commented on server/command/user_leave_rotation.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+func (c *Command) leaveRotation(parameters []string) (md.MD, error) {
+	c.withFlagRotation()
```
It's difficult because it is special (positional/flag combination).

A question I would like to ask (myself) is why is the rotation name/ID convention is different from say, the tasks. I have no good answer to that, yet. Perhaps need to either stick with IDs everywhere, no names (that'd be not user-friendly after the first "archive"), or come up with a better "unique active name" resolution (and flags) that works for all.

levb replied:
Inclined not to fix this, yet - another time.

levb commented on server/command/user_show.go:
```diff
@@ -4,23 +4,27 @@
+	if users.Len() == 1 {
+		return md.JSONBlock(users.AsArray()[0]), nil
+	}
+	return md.JSONBlock(users), nil
```
This one is a dinosaur from the previous implementation, will fix. Really, the convention should be `show` == 1, `list` == multiple.

levb replied:
1/5 Inclined to leave as is for now, will need to change in a round of more user-friendly command output anyway.

levb commented on server/plugin/plugin_api.go:
```diff
@@ -6,16 +6,19 @@ package plugin
 
 // This file provides an implementation of sl.PluginAPI
 var _ sl.PluginAPI = (*Plugin)(nil)
+var _ config.Store = (*Plugin)(nil)
```
I am considering the config commands to be `demo hacks` at this point.

If it's a mainstream capability, should define it within the `PluginAPI`, IMO, if/when it's mature.

levb commented on server/sl/sl_setup.go:
```diff
@@ -0,0 +1,162 @@
+	}
+}
+
+func withExpandedTask(idref *types.ID, task *Task) func(sl *sl) error {
```
Yes. The read-only keys are passed as pointers, so that the `with...` calls can use previously-initialized structs as dependencies, as in 
```go
  task := NewTask()
  rotation := NewRotation()
  err := sl.Setup(
    withLoadTask(taskID, &task),
    withLoadRotation(task.RotationID, &rotation),
  )
```
The example ^^ reads well, but does not work. `task.RotationID` value is used as of compile-time ("") which is not the intention. To use the actual task.RotationID as it is _after_ withLoadTask I chose to pass these read-only parameters by reference.

levb commented on server/sl/api_task_assign.go:
```diff
@@ -0,0 +1,55 @@
+	"github.com/mattermost/mattermost-plugin-solar-lottery/server/utils/types"
+)
+
+type InAssignTask struct {
```
Maybe at some point in the future, I can see more consolidation of these structs coming; Conversely, for documentation and strictness it _may_ make sense to define In/Out separately for each API.

levb commented on server/sl/filler/queue/task_filler.go:
```diff
@@ -0,0 +1,25 @@
+	return &taskFiller{}
+}
+
+func (*taskFiller) FillTask(r *sl.Rotation, task *sl.Task, logger bot.Logger) (*sl.Users, error) {
```
There's already a big ticket for this; it's present in the repo now only for illustration purposes.

mattermod commented:
The file `.circleci/config.yml` is in the blacklist and should not be modified from external contributors, please if you are part of the Mattermost Org submit this PR in the upstream.
 /cc @mattermost/core-security @mattermost/core-build-engineers

mattermod commented:
The file `.circleci/config.yml` is in the blacklist and should not be modified from external contributors, please if you are part of the Mattermost Org submit this PR in the upstream.
 /cc @mattermost/core-security @mattermost/core-build-engineers

levb commented on server/plugin/plugin_api.go:
```diff
@@ -67,16 +69,31 @@ func (p *Plugin) Clean() error {
 func (p *Plugin) SendEphemeralPost(channelID, userID, message string) {
 	ephemeralPost := &model.Post{
```
This is what p.bot.Ephemeral uses.

levb commented on server/plugin/plugin.go:
```diff
@@ -22,88 +17,68 @@ import (
+	config config.Service
 
-	Templates map[string]*template.Template
+	botUserID string
```
There are some issues because `p.botUserID` is cached once, in `OnActivate` but the bot is currently not a "service" and is freely instantiated in a few places. 1/5 suggest leaving as is, but refactoring later, making bot a "service".

levb commented on server/utils/kvstore/entity.go:
```diff
@@ -0,0 +1,70 @@
+	for i := 0; i < 5; i++ {
+		id := name
+		if i > 0 {
+			id = name + "-" + model.NewId()[:7]
```
Would you rather see it shorter?

levb commented on server/utils/types/time.go:
```diff
@@ -0,0 +1,112 @@
+// var _ json.Unmarshaler = (*Time)(nil)
+var _ pflag.Value = (*Time)(nil)
+
+func NewTime(tt ...time.Time) Time {
```
To make it optional, default being `time.NewTime.UTC()`

levb replied:
I am changing it in the next iteration; will be explicit (or I may kill this type after all for all use-cases except `pflag`)

levb commented:
@iomodo @larkox I've addressed most of the feedback, and took the liberty of resolving the comments that seemed resolved - just to keep the PR manageable.

1/5 I am inclined to merge this without the full parity with the prior code, and iterate on top of it. The missing functionality

levb commented:
@iomodo @larkox I added the `/lotto task unassign` command to fully merge master, including https://github.com/mattermost/mattermost-plugin-solar-lottery/pull/25. Please confirm and I'll merge,

Apologies, @TheoVitkovskiy this effectively overwrites your commit due to a large refactoring, but I appreciate the contribution!
